include/deal.II/matrix_free/cuda_tensor_product_kernels.cuh

```diff
@@ -0,0 +1,395 @@
+
+      /**
+       * Helper function for values() and gradients().
+       */
+      template <int direction, bool dof_to_quad, bool add, bool in_place>
+      __device__ void apply(Number shape_data[],
+                            const Number *in,
+                            Number       *out) const;
+
+      /**
+       * Evaluate the finite element function at the quadrature points.
+       */
+      __device__ void value_at_quad_pts(Number *u);
+
+      /**
+       * Helper function for integrate(). Integrate the finite element function.
+       */
+      __device__ void integrate_value(Number *u);
+
+      /**
+       * Evaluate the gradients of the finite element function at the quadrature
+       * points.
+       */
+      __device__ void gradient_at_quad_pts(const Number *const u,
+                                           Number *grad_u[dim]);
+
+      /**
+       * Helper function for integrate(). Integrate the gradients of the finite
+       * element function.
+       */
+      template <bool add>
+      __device__ void integrate_gradient(Number *u,
+                                         Number *grad_u[dim]);
+    };
+
+
+
+    template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+    __device__ EvaluatorTensorProduct<evaluate_general, dim, fe_degree,
+               n_q_points_1d, Number>::EvaluatorTensorProduct()
+    {}
+
+
+
+    template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+    template <int direction, bool dof_to_quad, bool add, bool in_place>
+    __device__ void EvaluatorTensorProduct<evaluate_general, dim, fe_degree,
+               n_q_points_1d, Number>::values(const Number *in,
+                                              Number *out) const
+    {
+      apply<direction, dof_to_quad, add, in_place>(global_shape_values, in, out);
+    }
+
+
+
+    template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+    template <int direction, bool dof_to_quad, bool add, bool in_place>
+    __device__ void EvaluatorTensorProduct<evaluate_general, dim, fe_degree,
+               n_q_points_1d, Number>::gradients(const Number *in,
+                                                 Number *out) const
+    {
+      apply<direction, dof_to_quad, add, in_place>(global_shape_gradients, in, out);
+    }
+
+
+
+    template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+    template <int direction, bool dof_to_quad, bool add, bool in_place>
+    __device__ void EvaluatorTensorProduct<evaluate_general, dim, fe_degree,
+               n_q_points_1d, Number>::apply(Number shape_data[],
+                                             const Number *in,
+                                             Number       *out) const
+    {
+      const unsigned int i = (dim == 1) ? 0 : threadIdx.x%n_q_points_1d;
+      const unsigned int j = (dim == 3) ? threadIdx.y : 0;
+      const unsigned int q =
+        (dim == 1) ? (threadIdx.x%n_q_points_1d) :
```

Comment:
This is one completely SIMD integer division, which is really negligible for a bandwidth-bound algorithm. It is there to let each CUDA block treat several cells, since they would otherwise be too small for low order. That only leads to different data though, and no branching. Or is there something else that I don't see?